### ISSUE TITLE
Memory Leak/Unbounded growth when Allocating descriptor sets/resetting descriptor pools

### ISSUE THREAD

**Issue by gununes** [KhronosGroup/Vulkan-ValidationLayers#2512]
**Describe the Issue**
Have a simple demo app where it allocates descriptors sets every frame and reset the descriptor pools associated with it. When validation layers are enabled, I see unbounded memory growth coming from the validation layers.dll. After 4 minutes running the app I see a growth of about ~40MB. The same behavior does not happen when disabling the validation layer. 

The repro steps are exactly the same of this issue: [https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/236](#236). Which is marked as fixed in 2018.

I have attached the callstack where the unbounded growth is coming from, it is on a concurrent_unordered_map::insert_or_assign call inside vulkan_layer_chassis::AllocateDescriptorSets. As mentioned above, in 4 minutes, this map has grown by ~40MB, and doesn't seem to stop.

**Environment:**
 - OS: Windows
 - GPU: Nvidia 2080RTX
 - SDK or header version if building from repo: Vulkan SDK 1.2.162.0

**Additional context**
See call stack below
![Capture](https://user-images.githubusercontent.com/44151632/106401203-b1616280-63d7-11eb-8391-6a68b4e9bebf.PNG)

**Comment by mark-lunarg** [comment]
Triage notes:  Thread safety was modified to track whether allocated descriptor sets and layouts used `UPDATE_AFTER_BIND`.  There are two maps, `dsl_update_after_bind_map `and `ds_update_after_bind_map `which are connected up to the DS allocate logic, but not to the reset/free logic.  The thread-safety manually written source code in `inline_custom_source_preamble `likely needs to be updated to clean up those maps when the sets/layouts are destroyed.

**Comment by nisse-work** [comment]
I'm also seeing various complaints about leaks, when I enable validation layers in application tests that are run with leak checks enabled. This can be worked around, by disabling leak checking around certain vulkan calls (using https://github.com/abseil/abseil-cpp/blob/2e94e5b6e152df9fa9c2fe8c1b96e1393973d32c/absl/debugging/leak_check.h#L108). But it would be desirable to fix the leaks properly.

I've made a branch at https://github.com/nisse-work/Vulkan-ValidationLayers/tree/enable-gcc-leak-sanitizer, that enables a run with gcc's built in leak checker, -fsanitize=leak, to the tests run by the CI. It fails pretty hard.

I would really appreciate work to fix leaks and enable checking as part of the CI. I'm having some trouble to contribute, though, since my local machine (gnu/linux) uses Intel Mesa drivers and there are lots of unrelated ways that tests fail. I do have a different machine using (proprietary) nvidia drivers, is the testsuite expected to work better there?

**Comment by ncesario-lunarg** [comment]
Thanks @nisse-work, we are working on getting nightly runs with `-fsanitize=address` turned on which should help with this.

> I do have a different machine using (proprietary) nvidia drivers, is the testsuite expected to work better there?

Yes, linux nvidia proprietary drivers should _work_ (I think we might have 1 or 2 tests black listed, but it's not nearly as bad as Intel + Mesa, for which we essentially don't check at the moment).

**Comment by nisse-work** [comment]
Hi, I have tests working (on intel mesa), with some minor fixes from PR #3013. To start somewhere, and not get a too huge report, I made a run of ./tests/vk_layer_validation_tests --gtest_filter='VkPositiveLayerTest.*' in a build with leak checking on.

I see a couple of leaks, stack traces including:

VkPositiveLayerTest_DestroySwapchainWithBoundImages_Test, line number pointing to

        vk::CreateImage(m_device->device(), &image_create_info, NULL, &image);

VkPositiveLayerTest_SeparateDepthStencilSubresourceLayout_Test, lines

    vkCreateRenderPass2KHR(device(), &rp2, nullptr, &render_pass_separate);
and
    vkCreateRenderPass2KHR(device(), &rp2, nullptr, &render_pass_combined);

A couple of leaks involving VkRenderFramework::InitSurface and XOpenDisplay (some fixes were commited recently, but reverted in b613b4a69ba96d6ba9b6fd3ceddb17e7a4a14d60, so I'm not reporting details here).

VkPositiveLayerTest_SubresourceLayout_Test, leak related to submit, line in test

    m_commandBuffer->QueueCommandBuffer();

and stack trace

    #18 0x560ecba70d69 in vk_testing::Queue::submit(std::vector<vk_testing::CommandBuffer const*, std::allocator<vk_testing::CommandBuffer const*> > const&, vk_testing::Fence const&, bool) /home/nisse/hack/Vulkan-ValidationLayers/tests/vktestbinding.cpp:405
    #19 0x560ecba70e34 in vk_testing::Queue::submit(vk_testing::CommandBuffer const&, vk_testing::Fence const&, bool) /home/nisse/hack/Vulkan-ValidationLayers/tests/vktestbinding.cpp:411
    #20 0x560ecba5c22a in VkCommandBufferObj::QueueCommandBuffer(vk_testing::Fence const&, bool) /home/nisse/hack/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:2258
    #21 0x560ecba5c1aa in VkCommandBufferObj::QueueCommandBuffer(bool) /home/nisse/hack/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:2252
    #22 0x560ecb9e5e65 in VkPositiveLayerTest_SubresourceLayout_Test::TestBody() /home/nisse/hack/Vulkan-ValidationLayers/tests/vkpositivelayertests.cpp:10921

VkPositiveLayerTest_SubresourceLayout_Test (same test) also line 

    vk::CmdPipelineBarrier(m_commandBuffer->handle(), VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, 0, 0,
                           nullptr, 0, nullptr, 1, &barrier);

and the other nearby calls to vk::CmdPipelineBarrier.

To sum up, most of these tests seem to work fine, and a fairly small number trigger leaks.

**Comment by nisse-work** [comment]
An update here: I've checked out the latest version, and with a plain build, tests now work on this laptop of mine (debian gnu/linux, intel graphics). But I then tried a configuring with the new option -DVVL_ENABLE_ASAN=ON.
```
cmake .. -C ../external/helper.cmake -DCMAKE_BUILD_TYPE=Debug -DBUILD_WSI_WAYLAND_SUPPORT=OFF -DUSE_ROBIN_HOOD_HASHING=OFF -DUPDATE_DEPS=ON -DBUILD_TESTS=ON -DVVL_ENABLE_ASAN=ON
```
Now running the same set of tests 
```
VK_LAYER_PATH=layers ./tests/vk_layer_validation_tests --gtest_filter='VkPositiveLayerTest.*'
```
results in
```
SUMMARY: AddressSanitizer: 342996 byte(s) leaked in 3962 allocation(s).
```

Unfortunately, backtraces I get from asan only says which test triggered each leak, stackframes within the validation layers are all displayed as just <unknown module>.

**Comment by jeremyg-lunarg** [comment]
In #3752 @moset pointed out you can get good stack traces by doing something like this in a leaking test case:
```
  // To get a proper stacktrace from ASAN. Maybe there's a better way. :)
  dlopen("libVkLayer_khronos_validation.so", RTLD_NOW);

```
The problem is that the Vulkan loader unloads the validation layer before asan translates back traces, and that dlopen() forces it to stay open. I agree that there's a better way or at least a way to hide this trick in our test framework when ASAN is enabled.

**Comment by jeremyg-lunarg** [comment]
@nisse-work, you may want to try again with PR #3754.

**Comment by nisse-work** [comment]
I've pulled latest changes, including that fix, and I can confirm it's an improvement. It fixes the leak in the simplest test which I've been looking into (VkPositiveLayerTest.EmptyDescriptorUpdateTest, with similar behavior as in Oskar's test cases).

Running all of VkPositiveLayerTest.* results in 

`SUMMARY: AddressSanitizer: 9496 byte(s) leaked in 18 allocation(s).`

(down from 3962 leaks before the fix). I tried adding the dlopen call to main(), but  I still get no useful stack traces. Perhaps I should try the main (non-asan) build and valgrind.
 
Running *all* tests still results in a large number of leaks, 

`SUMMARY: AddressSanitizer: 10822600 byte(s) leaked in 2844 allocation(s).`

**Comment by nisse-work** [comment]
Attaching the leaks reported by 

`VK_LAYER_PATH=layers valgrind --leak-check=full ./tests/vk_layer_validation_tests --gtest_filter='VkPositiveLayerTest.*'`

I can't really say if they're all in the validation layers, or if some are actually in the vulkan drivers.

[valgrind-leaks.txt](https://github.com/KhronosGroup/Vulkan-ValidationLayers/files/8060800/valgrind-leaks.txt)

Valgrind also produced a warning for a non-leak issue:

==1828894== Conditional jump or move depends on uninitialised value(s)
==1828894==    at 0x16ABA6FA: ObjectLifetimes::PreCallValidateQueueSubmit(VkQueue_T*, unsigned int, VkSubmitInfo const*, VkFence_T*) const (object_tracker.cpp:236)
==1828894==    by 0x16497E64: vulkan_layer_chassis::QueueSubmit(VkQueue_T*, unsigned int, VkSubmitInfo const*, VkFence_T*) (chassis.cpp:1147)
==1828894==    by 0x1FE3328: VkPositiveLayerTest_QueueSubmitSemaphoresAndLayoutTracking_Test::TestBody() (sync.cpp:368)
==1828894==    by 0x232B235: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2443)
==1828894==    by 0x2325A02: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2479)
==1828894==    by 0x23061B7: testing::Test::Run() (gtest.cc:2517)
==1828894==    by 0x2306AA4: testing::TestInfo::Run() (gtest.cc:2693)
==1828894==    by 0x23070E8: testing::TestCase::Run() (gtest.cc:2811)
==1828894==    by 0x2311DD0: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:5177)
==1828894==    by 0x232BEE2: bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest.cc:2443)
==1828894==    by 0x2326742: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest.cc:2479)
==1828894==    by 0x231099D: testing::UnitTest::Run() (gtest.cc:4786)
==1828894== 
[       OK ] VkPositiveLayerTest.QueueSubmitSemaphoresAndLayoutTracking (11816 ms)

Maybe I should file a separate issue for that.

**Comment by ziga-lunarg** [comment]
Closed by #3948
